treeherder/etl/taskcluster_pulse/handler.py

```diff
@@ -127,18 +129,14 @@ def handleMessage(message, taskDefinition=None):
         raise Exception("Unknown exchange: {exchange}".format(exchange=message["exchange"]))
     elif taskType == "pending":
         runId = message["payload"]["runId"]
-        # If the task run was created for an infrastructure rerun, then resolve
-        # the previous run as retried.
-        if runId > 0:
-            handleTaskRerun(parsedRoute, task, message["payload"])
-
+        # XXX: The original code used to send an extra Pulse message for the previous run to mark it as "retry"
         job = handleTaskPending(parsedRoute, task, message["payload"])
```

Comment:
@djmitche I just want your confirmation that removing this code makes sense to you as well.
`taskcluster-treeherder` in this case emits two Pulse messages: one for the current task + one for the previous *run* which was retried.

The message for the previous run (notice `message.runId-1`) would be ingested by Treeherder and would update the previous run as `retry`:

```python
async def handleTaskRerun(pushInfo, task, message):
    let job = this.buildMessage(pushInfo, task, message.runId-1, message);
    job = buildMessage(pushInfo, task, message["runId"]-1, message)
    job["state"] = "completed"
    job["result"] = "fail"
    job["isRetried"] = True   # <-- This is what makes Treeherder mark a job as `retried`
```

Now, all of these comments refer to this diff specifically (a compare with all related commits):
https://github.com/mozilla/treeherder/pull/5042/files/a23e935467dc7667ce7e76bc75a5c1338b6aa246..566d61bf334a44a6e9f24dd7eba3afd45b1b875e

We can know if an excepted task should be considered as retried by checking the  "reasonResolved":
https://github.com/mozilla/treeherder/pull/5042/files/a23e935467dc7667ce7e76bc75a5c1338b6aa246..566d61bf334a44a6e9f24dd7eba3afd45b1b875e#diff-609378e93352a69d9c31bfe55b0a8d64R254
```python
    # Based on https://github.com/taskcluster/taskcluster/blob/1a224a54c17fe7871c39b56e185268082623760f/services/queue/schemas/v1/task-exception-request.yml
    # and https://github.com/taskcluster/taskcluster/blob/f6ed2f3819e4cbaabc8c422852113803b103058f/services/queue/schemas/v1/task-status.yml#L81
    if jobRun["reasonResolved"] in ("worker-shutdown",
                                    "intermittent-task",
                                    "claim-expired"):
        job["isRetried"] = True
```
A problem with this is that we might have new states that would require a retry.

In the scenario where I ingested *all* tasks at once, my changes above seem to make sense, however, once we reach live ingestion I'm not too sure this will work.

I think I could keep the original logic by swapping send Pulse messages (`await this.publishJobMessage(pushInfo, job, message.status.taskId);`) for adding messages to the celery queue.

The other option would be to make `handleMessage` return a list of jobs that need to be processed (instead of just 1).

If you're indiferent as to how I solve it, let me know and I will figure it out. If you have a strong preference let me know.

Reply:
The old code would mark any run other than the most recent as `isRetried`.  While the reasons you identified ('worker-shutdown', etc.) can be a cause of a new run, it's also possible to get a new run with the `rerunTask` API method, even if the previous run was resolved as successful.  So I think the modification you've made would miss that case.  Also, there's a limit to the number of retries a task will do (usually 5), so a task resolved as `worker-shutdown` might *not* be retried if it that run was the last allowed run.

That said, I don't know if this is worth worrying about.  If "isRetried" indicates that a task was *automatically* retried, then arguably it shouldn't be set for `rerunTask` anyway.  But I think the full solution would be one of the two alternatives you mentioned.

Reply:
I have taken the approach of returning a list of jobs.
I have reverted these changes and we're now back to the same code logic we had earlier.

Reply:
I forgot to mention that I stripped the commits we were discussing.
Return list of jobs to be processed as Pulse messages is accomplished in here:
https://github.com/mozilla/treeherder/pull/5042/commits/a29ec9412188803755be5e22de452986dc9a038a